File: app/src/main/java/com/junkiedan/junkietuner/data/entities/Tuning.java
package com.junkiedan.junkietuner.data.entities;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

/**
 * Tuning entity that is used for the Room Database
 * @author Stavros Barousis
 */
@Entity
public class Tuning {

    // The unique ID of the tuning
    @PrimaryKey(autoGenerate=true)
    public int id;

    // Name of the tuning (For example "Standard E")
    @ColumnInfo(name = "name")
    public String name;

    // Notes string formatted as "[E2,A2,D3,G3,B3,E4]"
    @ColumnInfo(name = "notes")
    public String notes;

    public Tuning(String name, String notes) {
        this.name = name;
        this.notes = notes;
    }

    /**
     * Formatter function for the string that is displayed
     * on the user interface under the tunings fragment
     * @return Tuning in the UI format. For example the
     * "Standard E" Tuning "[E2,A2,D3,G3,B3,E4]" is returned as
     * "E2 A2 D3 G3 B3 E4" (between each notes there are 2
     * space characters.
     */
    public String notesFormatted() {
        StringBuilder sb = new StringBuilder();
        int len = notes.length();
        for (int i = 1; i < len - 1; ++i) {
            if (notes.charAt(i) == ',') {
                // Number of space characters that will be between
                // each note (Currently this value is set to 2.
                sb.append("  ");
            } else {
                sb.append(notes.charAt(i));
            }
        }
        return sb.toString();
    }

    @NonNull
    @Override
    public String toString() {
        return "Tuning{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", notes='" + notes + '\'' +
                '}';
    }
}

File: app/src/main/java/dev/thestbar/tunify/data/entities/Tuning.kt
package dev.thestbar.tunify.data.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity
data class Tuning(
    @JvmField @ColumnInfo(name = "name") var name: String,
    @JvmField @ColumnInfo(name = "notes") var notes: String
) {
    @JvmField
    @PrimaryKey(autoGenerate = true)
    var id: Int = 0

    fun notesFormatted(): String {
        val sb = StringBuilder()
        val len = notes.length
        for (i in 1 until len - 1) {
            if (notes[i] == ',') sb.append("  ") else sb.append(notes[i])
        }
        return sb.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Tuning) return false
        return id == other.id && name == other.name && notes == other.notes
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + name.hashCode()
        result = 31 * result + notes.hashCode()
        return result
    }
}
